Jl/beamsort

luojing1211 opened the pull request:
This is the new beamSort stage, with controllable dump size. Feel free to comment on it. An operation diagram is included. 
![BeamBufferSort](https://user-images.githubusercontent.com/5324931/120213464-3c752c80-c201-11eb-84e6-b790c67a8b9e.png)

Question: Since we have the beam sorted stage, do we need the mergeBeamBuffer stage?

luojing1211 committed:
add raw record config file

luojing1211 committed:
add raw record config file

luojing1211 committed:
Adding merge raw buffer stage

luojing1211 committed:
Adding merge raw buffer stage

luojing1211 committed:
add test config files.

luojing1211 committed:
add test config files.

luojing1211 committed:
use the memory indexing for the packet merge.

luojing1211 committed:
use the memory indexing for the packet merge.

luojing1211 committed:
add FredIDBeamMeta

luojing1211 committed:
add FredIDBeamMeta

luojing1211 committed:
Make buffer merge work

luojing1211 committed:
Make buffer merge work

luojing1211 committed:
Change config file to test the code

luojing1211 committed:
Change config file to test the code

luojing1211 committed:
change in_buf_turn to in_frame_counter

luojing1211 committed:
change in_buf_turn to in_frame_counter

luojing1211 committed:
add merged beam metadata

luojing1211 committed:
add merged beam metadata

luojing1211 committed:
add merged beam metadata

luojing1211 committed:
add merged beam metadata

luojing1211 committed:
add merged beam metadata to merged buffer

luojing1211 committed:
add merged beam metadata to merged buffer

luojing1211 committed:
update config file for beam packet merging

luojing1211 committed:
update config file for beam packet merging

luojing1211 committed:
add merged buffer split stage and its configure file

luojing1211 committed:
add merged buffer split stage and its configure file

luojing1211 committed:
Change the naming and fix the comments

luojing1211 committed:
Change the naming and fix the comments

luojing1211 committed:
add ndisk writing to merged buffer write

luojing1211 committed:
add ndisk writing to merged buffer write

luojing1211 committed:
fix n disk problem

luojing1211 committed:
fix n disk problem

luojing1211 committed:
Fix doc bugs

luojing1211 committed:
Fix doc bugs

luojing1211 committed:
add tracking beam recording to n disk config

luojing1211 committed:
add tracking beam recording to n disk config

luojing1211 committed:
Move the tracking beam test to tracking_beam_test folder

luojing1211 committed:
Move the tracking beam test to tracking_beam_test folder

luojing1211 committed:
move the tracking beam single record config

luojing1211 committed:
move the tracking beam single record config

luojing1211 committed:
update the config file for the full array recording

luojing1211 committed:
update the config file for the full array recording

luojing1211 committed:
make config accepte full chime nodes

luojing1211 committed:
make config accepte full chime nodes

luojing1211 committed:
Implement the feedbacks

luojing1211 committed:
Implement the feedbacks

luojing1211 committed:
Fix a doc bug

luojing1211 committed:
Fix a doc bug

luojing1211 committed:
Fix document bug

luojing1211 committed:
Fix document bug

luojing1211 committed:
use the old clang-format

luojing1211 committed:
use the old clang-format

luojing1211 committed:
move tracking beam to ndisk to main config

luojing1211 committed:
move tracking beam to ndisk to main config

luojing1211 committed:
remove tracking beam test config

luojing1211 committed:
remove tracking beam test config

luojing1211 committed:
add no metadata option to splitmergedbuffer

luojing1211 committed:
add no metadata option to splitmergedbuffer

luojing1211 committed:
Change notes

luojing1211 committed:
Change notes

luojing1211 committed:
add sort buffer stage

luojing1211 committed:
add sort buffer stage

luojing1211 committed:
add tracking beam data python reader

luojing1211 committed:
add tracking beam data python reader

luojing1211 committed:
add beambuffersort stage

luojing1211 committed:
add beambuffersort stage

luojing1211 committed:
add read tracking beam data reader.

luojing1211 committed:
add read tracking beam data reader.

luojing1211 committed:
add BeamBufferSort stage

luojing1211 committed:
add BeamBufferSort stage

luojing1211 committed:
add debug lines

luojing1211 committed:
add debug lines

luojing1211 committed:
add more infor to the merged meta

luojing1211 committed:
add more infor to the merged meta

luojing1211 committed:
Add more attributes to metadata

luojing1211 committed:
Add more attributes to metadata

luojing1211 committed:
add metadata

luojing1211 committed:
add metadata

luojing1211 committed:
add sorted recording config

luojing1211 committed:
add sorted recording config

luojing1211 committed:
mv file to test directory

luojing1211 committed:
mv file to test directory

luojing1211 committed:
untrack the test files.

luojing1211 committed:
untrack the test files.

luojing1211 committed:
remove old buffer sort code.

luojing1211 committed:
remove old buffer sort code.

luojing1211 committed:
Bring rawFileRead back

luojing1211 committed:
Bring rawFileRead back

luojing1211 committed:
Bring rawFileRead back to original

luojing1211 committed:
Bring rawFileRead back to original

luojing1211 committed:
Bring rawFileRead back to original

luojing1211 committed:
Bring rawFileRead back to original

luojing1211 committed:
start updating dump samples.

luojing1211 committed:
start updating dump samples.

luojing1211 committed:
add dump size

luojing1211 committed:
add dump size

luojing1211 committed:
enable different queue size

luojing1211 committed:
enable different queue size

luojing1211 committed:
update raw read

luojing1211 committed:
update raw read

luojing1211 committed:
update yaml file for different size

luojing1211 committed:
update yaml file for different size

luojing1211 committed:
update python reader

luojing1211 committed:
update python reader

luojing1211 committed:
Fix ctime in the header

luojing1211 committed:
Fix ctime in the header

luojing1211 committed:
add more checks

luojing1211 committed:
add more checks

luojing1211 committed:
add vdif stage

luojing1211 committed:
add vdif stage

luojing1211 committed:
add one disk vdif writing to the list

luojing1211 committed:
add one disk vdif writing to the list

james-s-willis commented on lib/stages/BeamBufferSort.cpp:
```diff
@@ -0,0 +1,414 @@
+            break;
+        
+	// Get the metadata for each frame.
+        if (has_freq_bin) {
+            // Frequency ID is in the metdata.
+            in_metadata = (FreqIDBeamMetadata*)get_metadata(in_buf, input_frame_id);
+            freq_bin = in_metadata -> frequency_bin;
+            fpga_seq_start = in_metadata -> fpga_seq_start;
+        }
+        else {
+	    // Frequency ID is not in the metadata. will add it to the metadata.
+            non_freq_metadata = (BeamMetadata*)get_metadata(in_buf, input_frame_id);
+            fpga_seq_start = non_freq_metadata -> fpga_seq_start;
+            const uint32_t num_freq_per_stream = Telescope::instance().num_freq_per_stream();
+            for (uint32_t f = 0; f < num_freq_per_stream; ++f) {
+                freq_bin = Telescope::instance().to_freq_id(non_freq_metadata -> stream_id, f);
+            }
+	    // Build the full Frequency IDed metadata
+	    in_metadata = temp_freq_metadata;
+	    nonfreq_meta_2_freq_meta(non_freq_metadata, in_metadata, freq_bin);
+        }
```
So sometimes the incoming beam metadata doesn't contain a frequency bin?

luojing1211 replied:
It needs a frequency mapping, but some other stage may add a frequency bin to it. For example the mergebeambuffer stage. So it would be better to have a check.

james-s-willis commented on lib/stages/BeamBufferSort.cpp:
```diff
@@ -0,0 +1,414 @@
+            fpga_seq_start = non_freq_metadata -> fpga_seq_start;
+            const uint32_t num_freq_per_stream = Telescope::instance().num_freq_per_stream();
+            for (uint32_t f = 0; f < num_freq_per_stream; ++f) {
+                freq_bin = Telescope::instance().to_freq_id(non_freq_metadata -> stream_id, f);
```
What happens when `num_freq_per_stream` is `> 1`? The value of `freq_bin` will get overwritten after each iteration of the loop.

luojing1211 replied:
That is something that needs to fix. If the num_freq_per_stream > 1. I don't know how the data is arranged. But for CHIME, this should be ok

james-s-willis commented on lib/stages/BeamBufferSort.cpp:
```diff
@@ -0,0 +1,414 @@
+	uint8_t* sub_frame_data = &sort_queue[time_offset][freq_bin][FreqIDBeamMeta_size];
+	memcpy(sub_frame_data, input, in_buf->frame_size);
+        // Debug line
+	FreqIDBeamMetadata* debug_meta = (FreqIDBeamMetadata*)&sort_queue[time_offset][freq_bin][0];
```
Isn't `debug_meta` equal to `sub_frame_metadata`? Also, should that `INFO` not be `DEBUG`?

james-s-willis reviewed (commented):
First round of comments.

james-s-willis commented on lib/stages/BeamBufferSort.cpp:
```diff
@@ -0,0 +1,414 @@
+	mark_frame_empty(in_buf, unique_name.c_str(), input_frame_id);
+        input_frame_id++;
+    }
+    // Free memory
+    // Free queue
+    for (int i = 0; i < wait_nframes; ++i){
+        for (uint32_t j = 0; j < total_freq_chan; j++){
+            free(sort_queue[i][j]);
+        }
+    // Free temp freqid metadata
+    free(temp_freq_metadata);
+    }
```
Can this not be done in the destructor?

luojing1211 replied:
temp_freq_metadata is defined in the `main_thread()`. I don't think it can be done in the destructor.

james-s-willis commented on lib/stages/BeamBufferSort.hpp:
```diff
@@ -0,0 +1,125 @@
+ * @par Buffers
+ * @buffer in_buf Kotekan single frame tracking beam buffer.
+ *     @buffer_format Array of @c chars
+ * @buffer out_buf Kotekan sorted tracking beam buffer.
```
Can describe how there can be multiple output buffers here.

james-s-willis commented on lib/stages/mergeBeamBuffer.cpp:
```diff
@@ -0,0 +1,138 @@
+        memcpy(sub_frame_data, in_frame, in_buf->frame_size);
+
+// Debug code. Check if the subframe matedata is the same.
+#ifdef DEBUGGING
```
Is the `DEBUGGING` macro defined by `kotekan`? Or do you define it yourself?

james-s-willis commented on lib/stages/nDiskFileWrite.hpp:
```diff
@@ -34,11 +34,13 @@
  * @conf num_disks      Int , the number of drives to read from
  * @conf disk_base      String, the path to the mounted drives
  * @conf disk_set       String, the disk name.
+ * @conf file_ext       String, Default 'vdif', the extenstion of the output file.
```
```suggestion
 * @conf file_ext       String, Default 'vdif', the extention of the output file.
```

james-s-willis commented on lib/stages/splitMergedBeamBuffer.cpp:
```diff
@@ -0,0 +1,128 @@
+	}
+	else{
+	    if (sub_frame_pre_frame == 0){
+	        ERROR("'sub_frame_pre_frame' is required in the configuration when the merged buffer metadata is not provided.");
```
I think you can just use `FATAL_ERROR` here without the need for `exit(-1)`.

james-s-willis commented on lib/stages/splitMergedBeamBuffer.cpp:
```diff
@@ -0,0 +1,128 @@
+                (FreqIDBeamMetadata*)&in_frame[sub_frame_start];
+	    DEBUG2("Frame start {:d}", sub_frame_start);
+            // Fill up the out put buffer metadata.
+            out_metadata->fpga_seq_start = sub_frame_metadata->fpga_seq_start;
```
Will a `memcpy` not work here?

james-s-willis reviewed (changes requested):
Do we have any unit tests in place for these changes? The logic in `BeamBufferSort` is quite complicated and could go wrong if not tested for edge cases. For `HFBAccumulate` I created a test that had missing frames to check that the integration could handle it. Could you do something similar? For example you could have the initial frame and miss `wait_nframes` frames and then add a frame with `fpga_seq_start` advanced in time. Another would be to receive only the same frequency over multiple time steps.

Also, are all config files needed? Or are some of them for testing and debugging purposes?

luojing1211 committed:
update vdif writing

luojing1211 committed:
update vdif writing

luojing1211 committed:
Fix vdif data order

luojing1211 committed:
Fix vdif data order

luojing1211 committed:
Fix bugs on ndisk output

luojing1211 committed:
Fix bugs on ndisk output

luojing1211 committed:
Change name

luojing1211 committed:
Change name

luojing1211 commented on config/beam_record_sorted.yaml:
```diff
@@ -0,0 +1,124 @@
+  # the size of the beam metadata with frequency is 72
+  frame_size: (dump_size * num_pol + freq_id_metadata_size) * sub_frames_per_merged_frame1
+  sorted_buffer_0:
+    kotekan_buffer: standard
+    #sorted_buffer_1:
+    #kotekan_buffer: standard
+    #sort_buffer_2:
+    #kotekan_buffer: standard
+    #sort_buffer_3:
+    #kotekan_buffer: standard
+    # sort_buffer_4:
+          # kotekan_buffer: standard
+    # sort_buffer_5:
+          # kotekan_buffer: standard
+    # sort_buffer_6:
+          # kotekan_buffer: standard
+    # sort_buffer_7:
+          # kotekan_buffer: standard
+    # sort_buffer_8:
+          # kotekan_buffer: standard
+    # sort_buffer_9:
+          # kotekan_buffer: standard
```
Those are for splitting frequencies to different disks. The number of buffers depends on the number of disks to use.

luojing1211 commented on lib/stages/BeamBufferSort.cpp:
```diff
@@ -0,0 +1,414 @@
+    // TODO this will be changed.
+    queue_frame_represent_size = samples_per_data_set; 
+    // Time delay vector for each channel
+    std::vector<double> time_delay(total_freq_chan, 0.0); // need to change here
```
This was put there for the gated pulsar. But I think I can remove it now

luojing1211 commented:
Most of the config files are for debugging purposes.  I will remove them before merging. Can we focus our review on the code first. After the code changes, I will have to test the new changes by using those config files.

luojing1211 committed:
Fix bug in the output buffer pointer

luojing1211 committed:
Fix bug in the output buffer pointer

luojing1211 committed:
update trachingbeambuffer

luojing1211 committed:
update trachingbeambuffer

luojing1211 committed:
update doc for beambuffersort

luojing1211 committed:
update doc for beambuffersort

luojing1211 commented:
Quick question, do we still want to keep the BeamBufferMerge stage? Since we are using the BeamBufferSort stage.

Jintao-Xu committed:
Add tracker stats in dump files (#989)

luojing1211 committed:
update docstrings

luojing1211 committed:
update docstrings

Jintao-Xu committed:
Add sidebar to show trackers_current info (#984)

- add current tracker info to sidebar
- show first two trackers in stage nodes

cubranic committed:
Use a `std::unordered_set` to keep track of frame ids to drop (#949)

tristpinsm committed:
Updates to pulsar gating script polyco-tools (#951)

* feat: add now option to set start and end times automatically.
* feat: add option to schedule pulsar gating update for later.
* fix: fix bad string formatting in scheduled command and seek to start of file for stdin.
* fix: update URLs in polyco-tools for coco.
* feat: catch user error end time earlier than start time.
* feat: add command for importing a 26m schedule file. (untested)
* feat: add option for phase offset from parfile.
* fix: strip phase wrappings from phase offset in polyco.
* feat: add option to use LST or UNIX time.
* fix: improve CLI help.
* fix: disable import_schedule.

Co-authored-by: Tristan Pinsonneault-Marotte <tristpinsm@phas.ubc.ca>
Co-authored-by: Richard Shaw <jrichardshaw@gmail.com>

Jintao-Xu committed:
Add dumped files to web interface (#987)

- give dump folder to debug_server.py with "-d=./dump"
- read dump files from the given folder
- generalize code from pipeline_view.js to use the dump data as well as current data
- Add slider to backtrace historical info in the dump files

Jintao-Xu committed:
Add cpu_ult to dump files and dump viewer (#993)

- move cpu trackers to kotekan trackers
- read cpu usage from /trackers_current
- show historical cpu usage in dump viewer

Jintao-Xu committed:
Update pipeline viewer doc (#994)

- update pipeline viewer doc

james-s-willis committed:
Label index in upchannelize kernel correctly (#999)

* Update upchannelize_flip_hfb.cl

andrerenard committed:
Improvements to the GPU patient objects (#986)

This PR is the first a few PRs designed to breakup #940 into more manageable PRs. 

* Adds a memory registration function for CUDA and OpenCL to improve host<->device async copy speeds
* Switch to using the new `KotekanTrackers` object to generate more stats from the command run times
* Add new profiling output functions for command line based kernel performance logging (useful for rapid testing)

andrerenard committed:
Add check to visAccumulate for number of output buffer frames (#997)

This change throws a runtime error if the condition in #992 occurs. To summarize that ticket: the system will get into a bad state if the number of frames in the output buffer of visAccumulate is less than num_freq_in_frame for multi-frequency configurations.

Closes #992

tristpinsm committed:
fix(visAccumulate): Fix frame age calculation. (#985)

andrerenard committed:
Configurable threshold for dropping frames in bufferSend (#1000)

andrerenard committed:
Merge pull request #1001 from kotekan/develop

Merge develop to master for 2021.10 release

andrerenard committed:
Fallback to system time if GPS time is not set (#1002)

Also fixes an issue where the ICETelescope object didn't return the gps_time_enabled() value correctly.

cubranic committed:
Send baseband to a receiver node instead of writing files directly (#927)

This change tears out the logic that existed to write baseband into files (i.e., over NFS) from the basebandReadout stage, and makes it truly just a "readout": when a baseband request arrives, the stage now copies out the requested data into an output Kotekan buffer, with event details recorded in the output frame metadata. The idea is that the output buffer will simply be forwarded to a baseband receiver node using a `bufferSend` stage.  Also adds multi-frequency support.

Co-authored-by: chitrang <chitrang.patel@mail.mcgill.ca>
Co-authored-by: Andre Renard <andre@renard.io>

andrerenard committed:
Merge branch 'master' of github.com:kotekan/kotekan into ar/merge_release_202110

andrerenard committed:
Merge pull request #1011 from kotekan/ar/merge_release_202110

Sync master release 2021.10 into develop

andrerenard committed:
fix(ICETelescope) Fix mapping for shuffle128 (#1004)

andrerenard committed:
fix(logging) Fix logging for OpenCL API errors (#1005)

andrerenard committed:
config(chime_keep_gpu_warm) Fix issues resulting from API changes (#1012)

* Adds the instrument coordinates
* Removes the baseband dump system config block (not longer needed with the new version).

andrerenard commented:
@luojing1211 Could you bring this branch upto data with develop, and fix the CI build and lint issues?  Let me know if you run into any issues with the linting (`iwyu` can be a tricky one to get working).

andrerenard reviewed (changes requested):
Can you reduce the number of config files to just the ones used in production?  I'd like to keep the active configuration files just to production and/or test cases with a pass/fall result in `config/tests/`.   Also which configuration file are you currently using for CHIME data capture?  T

andrerenard commented on lib/utils/util.c:
```diff
@@ -50,6 +50,27 @@ void make_rfi_dirs(int streamID, const char* write_to, const char* time_dir) {
+    if (err != -1) {
+        return err;
+    }
+    if (errno == EEXIST) {
```
We should adjust the logic here.  Do we care if the directory already exists?  Perhaps we could pass in a flag to indicate if we should fail if the directory exists or not?   At the very least if we want to exit the program regardless of if the failure was EEXIST or something else, we should log both cases.

Also it looks like a lot of the code in this file uses the old style logging.  We should use the `(INFO|WARN|ERROR)_F`, and `FATAL_ERROR_F()` macros instead.  At least in your new functions.

luojing1211 replied:
For the case of n disk vdif recording, it is always a new data directory that each time Kotekan starts. So we need to exit and throw an error at this place. Can you send me some examples of the new style of logging?

andrerenard commented on lib/utils/util.h:
```diff
@@ -40,6 +40,7 @@ extern "C" {
 #endif
 
 void make_rfi_dirs(int streamID, const char* write_to, const char* time_dir);
+int make_dir(const char* dir_name);
```
I'm not sure it makes sense to return the error code, if when we encounter an error we exit immediately?  If you want the caller to make this judgement (which is reasonable), then don't exit the program in the function.

luojing1211 replied:
Do you mean return an error code in the function, but the user can make a judgment in the program?

andrerenard commented on lib/stages/rawFileRead.cpp:
```diff
@@ -67,8 +67,13 @@ void rawFileRead::main_thread() {
         } else {
-            snprintf(full_path, full_path_len, "%s/%s_%07d.%s", base_dir.c_str(), file_name.c_str(),
-                     file_num, file_ext.c_str());
+            if (ndisk_wrote)
```
Why don't you loop over the num_disks?  If you are reading nDisk written files, don't you want to read from each disk?   I would also change this variable name to `ndisk_mode`.

luojing1211 replied:
This function was designed for different name conventions, during the development. I did not mean to change this stage. But somehow I have to commit it. Do you want me to change the stage?

andrerenard commented on lib/stages/CMakeLists.txt:
```diff
@@ -48,6 +51,8 @@ add_library(
     bufferDelay.cpp
     BufferSplit.cpp
     InputSubset.cpp
+    mergeBeamBuffer.cpp
+    BeamBufferSort.cpp
```
I'm also really bad at this, but we should try to be consistent with naming new stages with a starting capital letter.   (It will make it easier when we go back in and rename all the old lower case stages).

luojing1211 replied:
So it would be "MergeBeamBuffer.cpp" instead of "mergeBeamBuffer.cpp"?

andrerenard commented on lib/stages/oneDiskVDIFWrite.hpp:
```diff
@@ -0,0 +1,107 @@
+ * @conf   ref_year                 Double. The reference year for VDIF header. It only accepts
+ *                                  the integer year or the half year (e.g. 2020 or 2020.5)
+ *
+ * @author Jing Santiago Luo
+ *
+ *
+ */
+
+
+class oneDiskVDIFWrite : public kotekan::Stage {
```
Remove extra spaces
```suggestion
 * @author Jing Santiago Luo
 */
class oneDiskVDIFWrite : public kotekan::Stage {
```

andrerenard commented on lib/stages/oneDiskVDIFWrite.hpp:
```diff
@@ -0,0 +1,107 @@
+ * @file
+ * @brief Write sorted beam buffer to disk
+ *  - oneDiskVDIFWrite : public kotekan::Stage
+ */
+
+
+#ifndef ONE_DISK_VDIF_WRITE
+#define ONE_DISK_VDIF_WRITE
+
+#include "BeamMetadata.hpp" // for BeamMetadata
+#include "Config.hpp"       // for Config
+#include "Stage.hpp"
+#include "bufferContainer.hpp" // for bufferContainer
+#include "vdif_functions.h"   // for VDIFHeader
+
+
+#include <stdint.h> // for uint32_t
+#include <string>   // for string
+#include <vector>   // for vector
+
+
+using std::vector;
+
+
+/**
```
Remove extra spaces
```suggestion
 */

#ifndef ONE_DISK_VDIF_WRITE
#define ONE_DISK_VDIF_WRITE

#include "BeamMetadata.hpp" // for BeamMetadata
#include "Config.hpp"       // for Config
#include "Stage.hpp"
#include "bufferContainer.hpp" // for bufferContainer
#include "vdif_functions.h"   // for VDIFHeader

#include <stdint.h> // for uint32_t
#include <string>   // for string
#include <vector>   // for vector

using std::vector;

/**
```

andrerenard commented on lib/stages/oneDiskVDIFWrite.cpp:
```diff
@@ -0,0 +1,338 @@
+    vdif_frame_data_size = vdif_samples_per_frame * vdif_freq_per_frame * num_pol;
+    vdif_frame_size = vdif_frame_header_size + vdif_frame_data_size;
+    INFO("VDIF size {:d} {:d}", vdif_frame_data_size, vdif_frame_size);
+    vdif_frame = (uint8_t*) malloc(vdif_frame_size);
```
Check that `malloc()` returned valid memory, e.g. `!= nullptr`

andrerenard commented on lib/stages/oneDiskVDIFWrite.cpp:
```diff
@@ -0,0 +1,338 @@
+	        for (uint32_t jj = 0; jj < num_pol; jj++){
+	            //if (ii >= 0 && ii < 20){		
+	            //    if (tt > 620 && tt < 630 && time_offset > 49000 && time_offset < 51000)
+	   	    //        INFO("target_index {:d} src_index {:d} tt {:d} ii {:d} freq offset {:d} time offset {:d} frameszie {:d}", target_start +jj, src_start + jj, tt, ii, freq_offset, time_offset, in_buf_frame_size);
```
You should either make these commented out `INFO` statements into `DEBUG` or `DEBUG2` statements, or remove them entirely.

andrerenard commented on lib/stages/oneDiskVDIFWrite.cpp:
```diff
@@ -0,0 +1,338 @@
+    // cast vdif payload
+    payload = &vdif_frame[vdif_frame_header_size];
+    // Fill fixed parts of vdif frame header
+    vdif_header -> invalid = 0;
+    vdif_header -> legacy = 0;
+    // Only take the last two digits of the year.
+    // If the time beyond 2100, we should consider change the code here
+    vdif_header -> ref_epoch = (uint32_t)((ref_year - 2000) * 2) ; // Cast it to int. 
+    vdif_header -> unused = 0;
+    vdif_header -> log_num_chan = (int)log2 ((float)(vdif_freq_per_frame * num_pol));
+    vdif_header -> frame_len = (vdif_frame_data_size + vdif_frame_header_size) / 8;
+    vdif_header -> vdif_version = 0;
+    vdif_header -> data_type = 1;
+    vdif_header -> bits_depth = 4 - 1;  //  bits per sample minus 1
+    vdif_header -> station_id = 9999;
+    vdif_header -> edv = 0;   // for now
+    vdif_header -> eud1 = 0; 
+    vdif_header -> eud2 = 0;
+    vdif_header -> eud3 = 0;
+    vdif_header -> eud4 = 0;
```
The formatting here doesn't follow our design guide, e.g. no spaces in member access `vdif_header->invalid`.  I think once you run clang-format, that should automatically fix this.

andrerenard commented on lib/stages/oneDiskVDIFWrite.cpp:
```diff
@@ -0,0 +1,338 @@
+    INFO("VDIF size {:d} {:d}", vdif_frame_data_size, vdif_frame_size);
+    vdif_frame = (uint8_t*) malloc(vdif_frame_size);
+    // Cast vdif header
+    vdif_header = (VDIFHeader*) &vdif_frame[0];
```
Redundant array access and dereference.
```suggestion
    vdif_header = (VDIFHeader*) vdif_frame;
```

andrerenard commented on lib/stages/oneDiskVDIFWrite.cpp:
```diff
@@ -0,0 +1,338 @@
+    // Cast vdif header
+    vdif_header = (VDIFHeader*) &vdif_frame[0];
+    // cast vdif payload
+    payload = &vdif_frame[vdif_frame_header_size];
```
You could also use pointer math.  Although one could argue it's clearer to read with an array access.
```suggestion
    payload = vdif_frame + vdif_frame_header_size;
```

andrerenard commented on lib/stages/oneDiskVDIFWrite.cpp:
```diff
@@ -0,0 +1,338 @@
+    int payload_ctr;
+    uint8_t* input = nullptr;
+    frameID in_frame_id(in_buf);
+    uint8_t* vdif_frame;
```
It would be good practice to set all your pointers to `nullptr` when you declare them.

andrerenard reviewed (changes requested):
Not done with the review, but here are some comments to get started.

andrerenard commented:
Also take a look at the LGTM warnings, a few of those alerts look like they should be fixed.

luojing1211 committed:
Update lib/stages/oneDiskVDIFWrite.hpp

Co-authored-by: Andre Renard <andre@renard.io>

luojing1211 committed:
Delete empty space

luojing1211 committed:
Update lib/stages/oneDiskVDIFWrite.cpp

Co-authored-by: Andre Renard <andre@renard.io>
